### ISSUE TITLE
Misleading 'Timeout exceeded' for OPTIMIZE TABLE

### ISSUE THREAD

**Issue by 4ertus2** [ClickHouse/ClickHouse#6093]
It's not clear what really happened here.
19.11.3.11

```
OPTIMIZE TABLE <some_big_table> PARTITION 201902 FINAL

Timeout exceeded while receiving data from server. Waited for 300 seconds, timeout is 300 seconds.
Cancelling query.
Ok.
Query was cancelled.
```

**Comment by alexey-milovidov** [comment]
BTW, 300 seconds is default socket read timeout.

**Comment by bobelev** [comment]
Still an issue

```plain

OPTIMIZE TABLE default.notify_rewards FINAL

Timeout exceeded while receiving data from server. Waited for 300 seconds, timeout is 300 seconds.
Cancelling query.
Received exception from server (version 20.3.5):
Code: 999. DB::Exception: Received from localhost:9000. DB::Exception: Session expired, path: /clickhouse/tables/important/notify_rewards/replicas/ch814510/queue/queue-0007410722
.
Query was cancelled.
```

**Comment by alexey-milovidov** [comment]
It is totally possible to show a progress for OPTIMIZE query. At least when it is executed on the same server.

**Comment by alexey-milovidov** [comment]
Still relevant. The current behaviour is below the desired usability standard.

**Comment by mzealey** [comment]
I agree. It seems that tracking `system.merges` can give you an idea of progress at the moment.

**Comment by devicenull** [comment]
Any progress here?  Even making this return instantly would be better then the current behavior.

I'm trying to use optimize from a script (it massively reduces disk space usage on our setup), and I'm having to catch and ignore timeout errors here just to actually execute optimize.

**Comment by alexey-milovidov** [comment]
No progress so far.

**Comment by tisonkun** [comment]
> Misleading

Follow the thread still I don't know what happens actually.

**Comment by FedericoCeratto** [comment]
A practical workaround is to bump the timeout and run the query from the same command:
`clickhouse-client --receive_timeout 9999 --send_timeout 9999 -q 'OPTIMIZE TABLE ....'`

**Comment by lialsoftlab** [comment]
Got the same problem with the CH 23.3 -- but in my case timeout is set for 60 mins and OPTIMIZE TABLE FINAL getting job done in 10 .. 30 mins but very frequently the server doesn't responds to the client anyhow and app gets timeout error after 60 mins delay.

**Comment by gempir** [comment]
Anyone have this issue with http clients? 

I already have all timeouts increased that I can imagine.

```
http://127.0.0.1:48123?extremes=0&readonly=0&max_execution_time=3600&enable_http_compression=1&database=default&tcp_keep_alive_timeout=3600&send_timeout=3600&receive_timeout=3600
```

But my client dies every time with 

```
Operation timed out after 300000 milliseconds with 0 bytes received
```

**Comment by den-crane** [comment]
@gempir it's JDBC client / python / go settings. Has nothing to do with Clickhouse server.
https://stackoverflow.com/a/63622055/11644308

**Comment by gempir** [comment]
I'm running a third party php client.  https://github.com/smi2/phpClickHouse

I think jdbc/python/go mostly use the binary protocol, so I assume timeouts work very different there.

**Comment by gempir** [comment]
And the mentioned `socket_timeout` is indeed just something the clients use. Nothing from the server. 

I cannot set it in my client either. 

`Setting socket_timeout is neither a builtin setting nor started with the pr...`

**Comment by den-crane** [comment]
@gempir AFAIK `$db->setTimeout(300);`
https://github.com/search?q=repo%3Asmi2%2FphpClickHouse+setTimeout&type=code

**Comment by gempir** [comment]
@den-crane yeah already been down that road. That's why I'm asking here. I'm trying to understand the clickhouse server timeout settings, so maybe I can adjust the client to set the correct settings. 

I've set every setting I can think of 

```
        $this->clickhouseClient->setTimeout($timeoutSeconds);
        $this->clickhouseClient->setConnectTimeOut($timeoutSeconds);
        $this->clickhouseClient->settings()->max_execution_time($timeoutSeconds);
        $this->clickhouseClient->settings()->set('tcp_keep_alive_timeout', $timeoutSeconds);
        $this->clickhouseClient->settings()->set('send_timeout', $timeoutSeconds);
        $this->clickhouseClient->settings()->set('receive_timeout', $timeoutSeconds);
        $this->clickhouseClient->settings()->set('session_timeout', $timeoutSeconds);
        $this->clickhouseClient->write($query, $bindings);
        ```
        
        But the connection always dies at 300 seconds instead of my $timeoutSeconds = 3600

**Comment by den-crane** [comment]
@gempir you should ask in  https://github.com/smi2/phpClickHouse/ probable php libcurl has some max hardcoded settings for timeout.
